### PR TITLE
Clean FSP plugin and use a behavior similar to udevSR

### PR DESCRIPTION
**Pull request by MatiasVara**
I open this PR to discuss how I removed the workaround for the creation of vdis for the fsp vol driver.

### CODE REVIEW

**Comment by MatiasVara** [comment]
I have already submitted a PR at Runx repo. See at https://github.com/xcp-ng/runx/pull/1

**Comment by MatiasVara** [comment]
I think it could be nice if I add to this PR a couple of tests.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 1]
```diff
@@ -44,6 +49,7 @@ def create(self, dbg, sr_uuid, configuration, name, description):
             'name': name,
             'description': description,
             'uri': uri,
+            'vdis': {},
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: Why this change? You can store the VDIs and other info using the database.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 2]
```diff
@@ -60,6 +66,10 @@ def detach(self, dbg, sr):
         pass
 
     def ls(self, dbg, sr):
+        # clean metadata from broken paths
+        self._cleanvdis(dbg, sr)
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: `_clean_vdis` is probably a better name. :)

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 3]
```diff
@@ -92,6 +148,7 @@ def stat(self, dbg, sr):
             'uuid': meta['unique_id'],
             'overprovision': 0,
             'datasources': [],
+            'vdis': meta['vdis'],
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: Same, not sure it's a good idea to store this in the metadata.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 4]
```diff
@@ -70,6 +80,52 @@ def set_description(self, dbg, sr, new_description):
     def set_name(self, dbg, sr, new_name):
         util.update_sr_metadata(dbg, 'file://' + sr, {'name': new_name})
 
+    def _createvdi(self, sr, name, path, psize):
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: `_create_vdi` instead. :wink:

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 5]
```diff
@@ -70,6 +80,52 @@ def set_description(self, dbg, sr, new_description):
...
+        fsp = importlib.import_module("fsp")
+        cb = fsp.Callbacks()
+        description = "Created by sr.scan()"
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: I understand why you added this, but we also use an empty description. It's not mandatory.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py (hunk 4 above)
Comment (Wescoeur) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py]: Also you can remove the `psize` param.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 6]
```diff
@@ -70,6 +80,52 @@ def set_description(self, dbg, sr, new_description):
...
+                    volume = db.insert_new_volume(psize, image_type)
+                    db.insert_vdi(name, description, vdi_uuid, volume.id, sharable)
+                    volume_path = cb.volumeGetPath(opq, str(volume.id))
+            os.symlink(path, volume_path)
+        return vdi_uuid
+
+    def _scan(self, dbg, sr):
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        sr_dir = meta['uri'] + '/links'
+        log.debug('scanning: {}\n'.format(sr_dir))
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: I'm not sure it's useful to log here.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 7]
```diff
@@ -70,6 +80,52 @@ def set_description(self, dbg, sr, new_description):
...
+        if os.path.exists(sr_dir):
+            for filename in os.listdir(sr_dir):
+                path = os.path.join(sr_dir, filename)
+                if os.path.exists(path) and os.path.islink(path):
+                    if not path in meta['vdis']:
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: So after analysis, we can create a new table in the database like `vdi_custom_keys` instead of using meta data.
It's probably better to not store important data like other drivers in the JSON object.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 8]
```diff
@@ -70,6 +80,52 @@ def set_description(self, dbg, sr, new_description):
...
+    def _cleanvdis(self, dbg, sr):
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        new_vdis = {}
+        fsp = importlib.import_module("fsp")
+        cb = fsp.Callbacks()
+        for path in meta['vdis']:
+            with VolumeContext(cb, sr, 'w') as opq:
+                with PollLock(opq, 'gl', cb, 0.5):
+                    with cb.db_context(opq) as db:
+                        if os.path.exists(path) and db.get_vdi_by_id(meta['vdis'][path]) is not None:
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: Same here, we can fetch all VDIs using the database, then we can clean it. :slightly_smiling_face:

File: plugins/volume/org.xen.xapi.storage.fsp/volume.py [hunk 9]
```diff
@@ -11,61 +11,35 @@
...
+        # clean vdi metadata from sr
+        for path in meta['vdis']:
+            with VolumeContext(cb, sr, 'w') as opq:
+                with PollLock(opq, 'gl', cb, 0.5):
+                    with cb.db_context(opq) as db:
+                        if os.path.exists(path) and db.get_vdi_by_id(meta['vdis'][path]) is not None:
+                            log.debug('refreshing {} {}\n'.format(path, meta['vdis'][path]))
+                            new_vdis[path] = meta['vdis'][path]
+        meta['vdis'] = new_vdis
+        util.update_sr_metadata(dbg, 'file://' + sr, meta)
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/volume.py]: Normally, if the database is used, this code can be removed.

File: xapi/storage/libs/libcow/volume.py [hunk 10]
```diff
@@ -273,7 +273,10 @@ def stat(dbg, sr, key, cb):
                 _vdi_sanitize(vdi, opq, db, cb)
                 custom_keys = db.get_vdi_custom_keys(vdi.uuid)
 
-            psize = cb.volumeGetPhysSize(opq, str(vdi.volume.id))
+            psize = 0
+            if cb.volumeExists(opq, str(vdi.volume.id)):
+                psize = cb.volumeGetPhysSize(opq, str(vdi.volume.id))
+
```
Comment (Wescoeur) [inline xapi/storage/libs/libcow/volume.py]: IMHO, we should avoid this modification, this code is called in many drivers.
It's normal to raise an exception if the size cannot be fetched.

File: xapi/storage/libs/libcow/volume.py [hunk 11]
```diff
@@ -302,7 +305,10 @@ def ls(dbg, sr, cb):
 
                 image_format = ImageFormat.get_format(vdi.image_type)
 
-                psize = cb.volumeGetPhysSize(opq, str(vdi.volume.id))
+                psize = 0
+                if cb.volumeExists(opq, str(vdi.volume.id)):
+                    psize = cb.volumeGetPhysSize(opq, str(vdi.volume.id))
+
```
Comment (Wescoeur) [inline xapi/storage/libs/libcow/volume.py]: Same.

File: xapi/storage/libs/libcow/volume.py (hunk 10 above)
Comment (MatiasVara) [reply xapi/storage/libs/libcow/volume.py]: I need to double-check this change of behavior. When triggering `sr.scan()`, I observed that if the volume does not exist `sr.ls()` crashes without calling `sr.stat()`. I had somehow to prevent that crash otherwise `sr.scan()` keeps calling `sr.ls()` and crashing.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py (hunk 7 above)
Comment (MatiasVara) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py]: If I use `vdi_custom_keys` table, I could:
1. Get all custom keys by using `get_all_vdi_custom_keys()`
2. Iterate over the list to look for the key `path_to_link` and then check if the `path` is in use
3. If `path` is not in use, I create the vdi
    3.1. I set the new key for the new vdi uuid by using `set_vdi_custom_key()`

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py (hunk 8 above)
Comment (MatiasVara) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py]: By using the db, I shall:
1. Get all custom keys by using `get_all_vdi_custom_keys()`
2. For those that the key is `path_to_link` and the value does not exist, I call `delete_vdi_custom_key()`. If I understand well, when a vdi is deleted, all the `vdi_custom_keys` are deleted from the table. So this clean-up works only when the user intentionally removes a link. I will come back to this use-case later because I am not sure yet if it is a use-case that would like to support.

File: plugins/volume/org.xen.xapi.storage.fsp/volume.py (hunk 9 above)
Comment (MatiasVara) [reply plugins/volume/org.xen.xapi.storage.fsp/volume.py]: Got it. If the db is used, the `vdi_delete` removes all the `custom_keys` and this is not needed anymore.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py (hunk 1 above)
Comment (MatiasVara) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py]: This has been removed.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py (hunk 3 above)
Comment (MatiasVara) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py]: This has been removed.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py (hunk 8 above)
Comment (MatiasVara) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py]: I have removed _clean_vdis() since the list is cleaned during vdi.destroy().

File: xapi/storage/libs/libcow/volume.py (hunk 10 above)
Comment (MatiasVara) [reply xapi/storage/libs/libcow/volume.py]: I moved stat to the driver.

File: xapi/storage/libs/libcow/volume.py (hunk 11 above)
Comment (MatiasVara) [reply xapi/storage/libs/libcow/volume.py]: I moved stat to the volume driver.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 12]
```diff
@@ -70,6 +76,50 @@ def set_description(self, dbg, sr, new_description):
...
+        all_paths = self._get_paths_to_link(sr)
+        if os.path.exists(sr_dir):
+            for filename in os.listdir(sr_dir):
+                path = os.path.join(sr_dir, filename)
+                if os.path.exists(path) and os.path.islink(path):
+                    if not path in all_paths:
+                        vdi_uuid = self._create_vdi(sr, filename, path)
+                        log.debug('New VDI for symlink {}: uuid={}'.format(filename, vdi_uuid))
+
     def stat(self, dbg, sr):
```
Comment (MatiasVara) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py:145]: We shall add a log during either stat or ls to tell the user that a symlink is broken. The user has to issue a vdi_destroy to stop such a log.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 13]
```diff
@@ -60,8 +65,41 @@ def detach(self, dbg, sr):
...
+                    'description': vdi.description,
+                    'read_write': True,
+                    'virtual_size': vdi.volume.vsize,
+                    'physical_utilisation': 0,
+                    'uri': [image_format.uri_prefix + vdi_uri],
+                    'keys': custom_keys,
+                    'sharable': bool(vdi.sharable)
+                })
+
+        return results
```
Comment (MatiasVara) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: I moved the SR.ls() from the COWVolume to the SR driver due to psize. This is symmetric with SR.stat().

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 14]
```diff
@@ -60,8 +64,41 @@ def detach(self, dbg, sr):
         pass
 
     def ls(self, dbg, sr):
-        return COWVolume.ls(
-            dbg, sr, importlib.import_module('fsp').Callbacks())
+        # check for broken symlinks
+        self._check_paths(sr)
```
Comment (MatiasVara) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: I check if any symlink is broken and, in that case, write a debug log.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py (hunk 13 above)
Comment (Wescoeur) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py]: You can remove all this code using this piece instead:
```python
def ls(self, dbg, sr):
    self._check_paths(sr)
    self._scan(dbg, sr)
    return COWVolume.ls(dbg, sr, fsp.Callbacks())
```
And with a `fsp.py` file:

```python
import xapi.storage.libs.libcow.callbacks

class Callbacks(xapi.storage.libs.libcow.callbacks.Callbacks):
    def getVolumeUriPrefix(self, opq):
        return "fsp/" + opq + "|"
    
    # With this definition, the default `ls` method can call `volumeGetPhysSize` without error.
    # Otherwise an exception is thrown because the path is a folder instead of VHD or raw device.
    def volumeGetPhysSize(self, opq, name):
        return 0
```

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 15]
```diff
@@ -70,6 +107,63 @@ def set_description(self, dbg, sr, new_description):
...
+        vdi_uuid = str(uuid.uuid4())
+        with VolumeContext(cb, sr, 'w') as opq:
+            image_type = ImageFormat.IMAGE_DIRECTORY
+            image_format = ImageFormat.get_format(image_type)
+            with PollLock(opq, 'gl', cb, 0.5):
+                with cb.db_context(opq) as db:
+                    volume = db.insert_new_volume(0, image_type)
+                    db.insert_vdi(name, description, vdi_uuid, volume.id, sharable)
+                    volume_path = cb.volumeGetPath(opq, str(volume.id))
+                    db.set_vdi_custom_key(vdi_uuid, 'path_to_link', path, True)
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: I think we can use a global variable in this file instead of `'path_to_link'`. For example:
```python
DIRECTORY_KEY = 'directory'
...
db.set_vdi_custom_key(vdi_uuid, DIRECTORY_KEY, path, True)
```

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 16]
```diff
@@ -70,6 +107,63 @@ def set_description(self, dbg, sr, new_description):
...
+        return vdi_uuid
+
+    def _check_paths(self, sr):
+        all_vdis = {}
+        fsp = importlib.import_module("fsp")
+        cb = fsp.Callbacks()
+        with VolumeContext(cb, sr, 'w') as opq:
+            with PollLock(opq, 'gl', cb, 0.5):
+                with cb.db_context(opq) as db:
+                    all_vdis = db.get_all_vdi_custom_keys(True)
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py:108]: Good. Just a simple remark, and it's not mandatory for this PR, but it can be interesting to use a new method to filter on the `DIRECTORY_KEY` key. So just add a comment like: 
```python
# TODO: Filter on `DIRECTORY_KEY` instead of fetching all keys.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py (hunk 15 above)
Comment (Wescoeur) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py]: After that it's necessary to change all `'path_to_link'` references.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 17]
```diff
@@ -70,6 +107,63 @@ def set_description(self, dbg, sr, new_description):
...
+        for row in all_vdis:
+            if "path_to_link" in all_vdis[row]:
+                if not os.path.exists(all_vdis[row]["path_to_link"]):
+                    log.debug('Broken symlink {}: uuid={}'.format(all_vdis[row]["path_to_link"], row))
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: You can use a temporary variable here:
```python
directory = all_vdis[row]["path_to_link"]
if not os.path.exists(directory):
    log.debug('Broken symlink {}: uuid={}'.format(directory, row))
```
`vdi_uuid` is also more explicit than `row`. :)

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 18]
```diff
@@ -70,6 +107,63 @@ def set_description(self, dbg, sr, new_description):
...
+
+    def _get_paths_to_link(self, sr):
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: IMHO, `_get_all_volume_directories` is probably more explicit. :wink:

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 19]
```diff
@@ -70,6 +107,63 @@ def set_description(self, dbg, sr, new_description):
...
+        all_vdis = {}
+        fsp = importlib.import_module("fsp")
+        cb = fsp.Callbacks()
+        with VolumeContext(cb, sr, 'w') as opq:
+            with PollLock(opq, 'gl', cb, 0.5):
+                with cb.db_context(opq) as db:
+                    all_vdis = db.get_all_vdi_custom_keys(True)
+        all_paths = []
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: You can use a dictionary here. It depends on the usage, but it can give a better scaling if a lot of volumes are used.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 20]
```diff
@@ -70,6 +107,63 @@ def set_description(self, dbg, sr, new_description):
...
+        for row in all_vdis:
+            if "path_to_link" in all_vdis[row]:
+                all_paths.append(all_vdis[row]["path_to_link"])
+        return all_paths
+
+    def _scan(self, dbg, sr):
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        sr_dir = meta['uri'] + '/links'
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: We can change the behavior in another PR, but you can add a TODO about removal of this subdirectory here.
Also I think we can rename `/links` to `/directories`, because there is no barrier to use directly a folder instead of a symlink with this driver.

File: plugins/volume/org.xen.xapi.storage.fsp/volume.py [hunk 21]
```diff
@@ -11,62 +11,51 @@
...
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        cb = self.callbacks
+        with VolumeContext(cb, sr, 'w') as opq:
+            with Lock(opq, 'gl', cb):
+                with cb.db_context(opq) as db:
+                    vdi = db.get_vdi_by_id(key)
+                    db.delete_vdi(key)
+                with cb.db_context(opq) as db:
+                    cb.volumeDestroy(opq, str(vdi.volume.id))
+                    db.delete_volume(vdi.volume.id)
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/volume.py]: You can remove this duplication calling the parent method with `super`.
And probably better: you can delete this full `destroy` impl, because I don't see specific new action here. :sweat_smile:

File: plugins/volume/org.xen.xapi.storage.fsp/volume.py [hunk 22]
```diff
@@ -11,62 +11,51 @@
...
 
-        shared_dir = os.path.normpath(urlparse.urlparse(name).path)
-        if not shared_dir or shared_dir == '.':
-            raise ValueError('shared_dir param is empty')
-        if not os.path.isdir(shared_dir):
-            raise ValueError('shared_dir param is not a valid directory')
+    def stat(self, dbg, sr, key):
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/volume.py]: Same here, can you use the default `stat` implementation?

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py (hunk 19 above)
Comment (MatiasVara) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py]: Could you elaborate pls?

Comment (Wescoeur) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py]:
```python
all_paths = set()
for row in all_vdis:
    if "path_to_link" in all_vdis[row]:
        all_paths.add(all_vdis[row]["path_to_link"])
        return all_paths
```

It's just a small improvement. :wink:

File: plugins/volume/org.xen.xapi.storage.fsp/volume.py [hunk 23]
```diff
@@ -3,68 +3,10 @@
 import importlib
 import os
 import sys
-import urlparse
-import uuid
 import xapi.storage.api.v5.volume
-
 from xapi.storage import log
 from xapi.storage.libs import util
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/volume.py]: This import is not used. ;)

File: plugins/volume/org.xen.xapi.storage.fsp/volume.py [hunk 24]
```diff
@@ -3,68 +3,10 @@
...
-            'virtual_size': psize,
-            'physical_utilisation': 0,
-            'uri': [image_format.uri_prefix + vdi_uri],
-            'sharable': sharable,
-            'keys': {}
-        }
+from xapi.storage.libs.libcow.volume_implementation import Implementation
 
 
 def call_volume_command():
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/volume.py:11]: Because the features `VDI_CREATE` and `VDI_DESTROY` are not available, we should remove this block:
```python
    if base == "Volume.create":
        cmd.create()
    elif base == "Volume.destroy":
        cmd.destroy()
```

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 25]
```diff
@@ -4,14 +4,20 @@
...
 from xapi.storage.libs.libcow.volume import COWVolume
+from xapi.storage.libs.libcow.callbacks import VolumeContext
+from xapi.storage.libs.libcow.imageformat import ImageFormat
+from xapi.storage.libs.libcow.lock import PollLock
+from xapi.storage.libs.libcow.volume import _vdi_sanitize
 import xapi.storage.api.v5.volume
 
 import importlib
 
+DIRECTORY_KEY = 'path_to_link'
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: Maybe `'directory'` string instead (we should support symlinks and directories, so a generic term is probably better).

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 26]
```diff
@@ -70,6 +80,66 @@ def set_description(self, dbg, sr, new_description):
...
+            with PollLock(opq, 'gl', cb, 0.5):
+                with cb.db_context(opq) as db:
+                    # TODO: Filter on `DIRECTORY_KEY` instead of fetching all keys.
+                    all_vdis = db.get_all_vdi_custom_keys(True)
+        for vdi_uuid in all_vdis:
+            if DIRECTORY_KEY in all_vdis[vdi_uuid]:
+                directory = all_vdis[vdi_uuid][DIRECTORY_KEY]
+                if not os.path.exists(directory):
+                    log.debug('Broken symlink {}: uuid={}'.format(directory, vdi_uuid))
+
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py:115]: Little performance improvement + log update:
```python
for vdi_uuid, vdi in all_vdis.iteritems():
    directory = vdi.get(DIRECTORY_KEY)
    if directory is not None and not os.path.exists(directory):
        log.debug('Broken symlink or missing directory {}: uuid={}'.format(directory, vdi_uuid))
```

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 27]
```diff
@@ -70,6 +80,66 @@ def set_description(self, dbg, sr, new_description):
...
+        fsp = importlib.import_module("fsp")
+        cb = fsp.Callbacks()
+        with VolumeContext(cb, sr, 'w') as opq:
+            with PollLock(opq, 'gl', cb, 0.5):
+                with cb.db_context(opq) as db:
+                    all_vdis = db.get_all_vdi_custom_keys(True)
+        all_paths = set()
+        for vdi_uuid in all_vdis:
+            if DIRECTORY_KEY in all_vdis[vdi_uuid]:
+                all_paths.add(all_vdis[vdi_uuid][DIRECTORY_KEY])
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: Same, just improvement.
```python
for vdi in all_vdis.itervalues():
    directory = vdi.get(DIRECTORY_KEY)
    if directory is not None:
        all_paths.add(directory)
```

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 28]
```diff
@@ -60,8 +66,12 @@ def detach(self, dbg, sr):
         pass
 
     def ls(self, dbg, sr):
-        return COWVolume.ls(
-            dbg, sr, importlib.import_module('fsp').Callbacks())
+        # check for broken symlinks
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: IMHO, the call below is pretty self-explanatory without this comment. :)

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py (hunk 20 above)
Comment (Wescoeur) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py]: The `directories` name is more correct with direct directory support. ;)

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 29]
```diff
@@ -70,6 +80,66 @@ def set_description(self, dbg, sr, new_description):
...
+
+    def _scan(self, dbg, sr):
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        # TODO: A future implementation should remove the links directory
+        sr_dir = meta['uri'] + '/links'
+        all_paths = self._get_all_volume_directories(sr)
+        if os.path.exists(sr_dir):
+            for filename in os.listdir(sr_dir):
+                path = os.path.join(sr_dir, filename)
+                if os.path.exists(path) and os.path.islink(path):
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: I think we can use this condition instead:
```python
    # It supports symlinks + directories.
    if os.path.isdir(path):
```

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 30]
```diff
@@ -70,6 +80,66 @@ def set_description(self, dbg, sr, new_description):
...
+                    if not path in all_paths:
+                        vdi_uuid = self._create_vdi(sr, filename, path)
+                        log.debug('New VDI for symlink {}: uuid={}'.format(filename, vdi_uuid))
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]:
```python
log.debug('New VDI with path {}: uuid={}'.format(filename, vdi_uuid))
```

**Review by Wescoeur — CHANGES_REQUESTED** [review]

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py (hunk 26 above)
Comment (MatiasVara) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py:115]: I am OK with the change, but I would like to understand why this is a performance improvement, is it not still O(n)? Or I am missing something in those built-in functions.

Comment (benjamreis) [reply plugins/volume/org.xen.xapi.storage.fsp/sr.py:115]: `.iteritems` returns both key and value of `all_vdis` which saves a get when getting the `directory`.
Also it's shorter and easier to read :)

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 31]
```diff
@@ -4,14 +4,21 @@
...
 import urlparse
+import uuid
 
 from xapi.storage import log
 from xapi.storage.libs import util
 from xapi.storage.libs.libcow.volume import COWVolume
+from xapi.storage.libs.libcow.callbacks import VolumeContext
+from xapi.storage.libs.libcow.imageformat import ImageFormat
+from xapi.storage.libs.libcow.lock import PollLock
+from xapi.storage.libs.libcow.volume import _vdi_sanitize
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: Not used.

File: plugins/volume/org.xen.xapi.storage.fsp/sr.py [hunk 32]
```diff
@@ -70,6 +80,69 @@ def set_description(self, dbg, sr, new_description):
...
+        all_paths = set()
+        for vdi in all_vdis.itervalues():
+            directory = vdi.get(DIRECTORY_KEY)
+            if directory is not None:
+                all_paths.add(directory)
+        return all_paths
+
+    def _scan(self, dbg, sr):
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        # TODO: A future implementation should remove the DIRECTORY_PATH directory
```
Comment (Wescoeur) [inline plugins/volume/org.xen.xapi.storage.fsp/sr.py]: `DIRECTORIES_PATH` ;)

**Review by Wescoeur — CHANGES_REQUESTED** [review]

**Review by benjamreis — APPROVED** [review]

**Review by Wescoeur — APPROVED** [review]
